iceoryx2/src/service/attribute.rs

```diff
@@ -99,59 +99,99 @@
+
+    use core::hash::Hash;
+    use core::hash::Hasher;
+
+    use iceoryx2_bb_container::semantic_string;
+    use iceoryx2_bb_container::semantic_string::SemanticString;
+    use iceoryx2_bb_derive_macros::ZeroCopySend;
+
+    use crate::constants::MAX_ATTRIBUTE_KEY_LENGTH;
+
+    semantic_string! {
+      /// Fixed string for service attribute keys.
+      name: FixedString,
+      capacity: MAX_ATTRIBUTE_KEY_LENGTH,
+      invalid_content: |string: &[u8]| {
+        // empty keys are not allowed to be empty
+        string.is_empty()
+      },
+      invalid_characters: |_string: &[u8]| {
+         false
+      },
+      normalize: |this: &FixedString| {
+          this.clone()
+      }
+    }
+}
 
-type AttributeKeyString = FixedSizeByteString<MAX_ATTRIBUTE_KEY_LENGTH>;
-type AttributeValueString = FixedSizeByteString<MAX_ATTRIBUTE_VALUE_LENGTH>;
+/// Module containing the value type used for service attributes.
+pub mod value {
+
+    use core::hash::Hash;
+    use core::hash::Hasher;
+
+    use iceoryx2_bb_container::semantic_string;
+    use iceoryx2_bb_container::semantic_string::SemanticString;
+    use iceoryx2_bb_derive_macros::ZeroCopySend;
+
+    use crate::constants::MAX_ATTRIBUTE_VALUE_LENGTH;
+
+    semantic_string! {
+      /// Fixed string for service attribute values.
+      name: FixedString,
+      capacity: MAX_ATTRIBUTE_VALUE_LENGTH,
+      invalid_content: |string: &[u8]| {
+        // empty keys are not allowed to be empty
+        string.is_empty()
+      },
+      invalid_characters: |_string: &[u8]| {
+         false
+      },
+      normalize: |this: &FixedString| {
+          this.clone()
+      }
+    }
+}
 
-type KeyStorage = FixedSizeVec<AttributeKeyString, MAX_ATTRIBUTES>;
+// TODO: move into module ...
+type KeyStorage = FixedSizeVec<key::FixedString, MAX_ATTRIBUTES>;
 type AttributeStorage = FixedSizeVec<Attribute, MAX_ATTRIBUTES>;
 
 /// Represents a single service attribute (key-value) pair that can be defined when the service
 /// is being created.
 #[derive(Debug, Eq, PartialEq, Clone, PartialOrd, Ord, ZeroCopySend, Serialize, Deserialize)]
 #[repr(C)]
 pub struct Attribute {
-    key: AttributeKeyString,
-    value: AttributeValueString,
-}
-
-/// Errors that can occur when creating an [`Attribute`].
-pub enum AttributeError {
-    /// The attribute key exceeds the maximum allowed length.
-    KeyTooLong,
-    /// The attribute value exceeds the maximum allowed length.
-    ValueTooLong,
+    key: key::FixedString,
+    value: value::FixedString,
 }
 
 impl Attribute {
     /// Creates an attribute instance
-    pub fn new(key: &str, value: &str) -> Result<Self, AttributeError> {
-        Ok(Self {
-            key: AttributeKeyString::try_from(key).map_err(|_| AttributeError::KeyTooLong)?,
-            value: AttributeValueString::try_from(value)
-                .map_err(|_| AttributeError::ValueTooLong)?,
-        })
+    pub fn new(key: key::FixedString, value: value::FixedString) -> Self {
```

Comment:
The parameters should be references. `pub fn new(key: &key::FixedString, value: &value::FixedString) -> Self {`

Reply:
@elfenpiff 

I purposely chose not to use references in all of these instances. The reasons:
1. In cases where the key or value are no longer needed by the caller, computation time (for a copy) is saved
2. In the cases where the key or value is needed for later reuse by the caller, they must explicitly call `clone()`, which makes the computation time required for the copy visible

@elBoberido Could you help break the stale mate here ? 😆

Reply:
@orecham AFAIK, moving is done via memcpy and since those types are fixed size instead of just a pointer to some heap allocated memory, it might be more expensive to move them into the function instead of passing a reference and call clone manually. I'm not 100% sure, though.

cc @elfenpiff

Reply:
OK, so then we have:

1. When `AttributeKey` or `AttributeValue` are moved, it is potentially done using a memcpy (I did not consider this) which more computationally expensive than passing the reference and ensuring `clone` is used within the implementation
2. Using `clone()` has the potential to be more performant due to things like `PlacementDefault`

which means references at the API is better. I have updated the code.

@elfenpiff @elBoberido